<html>
	<head><title>GOBU - Plugin - Genome View Plugin</title>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-12601445-3']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
	</head>
	<table border=2 align=right><tr><td><a href="../index.html">last level</a></td></tr></table>
	<body>
		<h1><a name="top">Genome View Plugin</a></h1>

		<h3>Contents</h3>
		<ol>
		<li><a href="#description">Description</a>
		<li><a href="#usage">Usage by Example</a>
		<li><a href="#tips">Special Tips</a>
		<li><a href="#configTree">Config Tree</a>
		</ol>

		<table border=2 align=right><tr><td><a href="#top">top</a></td></tr></table>
		<h3><a name="description">Description</a></h3>
		This plugin will draw user selected R-nodes according their assigned location information (P-nodes representing LOC data type) on corresponding genome by specified colors (see Fig. 1). Enlarged views for interested chromosome regions are also provided (see Fig. 2).
		<table border=0 align=center>
			<tr>
				<td><a href="pl001001.png.html" target=_blank>
				<image src="pl001001t.png" alt="click to enlarge" align=right>
				</a></td>
				<td><a href="pl001002.png.html" target=_blank>
				<image src="pl001002t.png" alt="click to enlarge" align=right>
				</a></td>
			</tr>
			<tr align=center>
				<td><b>Fig. 1</b></td>
				<td><b>Fig. 2</b> The marked region in Fig. 1 is enlarged in this window.</td>
			</tr>
		</table>
		<br clear=all>

		<table border=2 align=right><tr><td><a href="#top">top</a></td></tr></table>
		<h3><a name="usage">Usage by Example</a></h3>
		<ol>
		<li>
		File -&gt; Load, open the file "HumanGene.tree"
		<br clear=all>
		
		<li>
		Plugins -&gt; Genome View. A window called "Genome View Control" will appear, usually in the upper-left corner of your screen.
		<table border=0 align=center>
			<tr><td><image src="pl001003.png" align=right></td></tr>
			<tr align=center><td><b>Fig. 3</b></td></tr>
		</table>
		<br clear=all>
		
		<li>
		Double-click on "Human", then a window called "GenomeView - Human" will appear. This window will display all human chromosomes. Use <i>Scale Down</i> and <i>Scale Up</i> buttons to adjust chromosomes to feasible sizes.
		<table border=0 align=center>
			<tr><td><a href="pl001004.png.html" target=_blank>
				<image src="pl001004t.png" alt="click to enlarge" align=right>
			</a></td></tr>
			<tr align=center><td><b>Fig. 4</b></td></tr>
		</table>
		<br clear=all>
		
		<li>
		Back to GOBU window, click user node "User Data", then all locations of human gene IDs are marked on chromosomes.
		<br clear=all>
		
		<li>
		Click user node "chr2", then all previously marked locations are removed and locations (i.e. LOC P-nodes) <i>under</i> chr2 are newly marked. Note that there are some marked locations not on chromosome 2, this might be resulted from non-synchronized data.
		<br clear=all>

		<li>
		Check <i>Additive</i> box (i.e. ensure this option is selected), select a new color in <i>Use color</i> list, then click user node "chr3". You will see chr2 and chr3 LOC P-nodes are marked by their assigned colors. chr2 LOC P-nodes will note be removed because checked <i>Additive</i> option ensures the additive drawing.
		<table border=0 align=center>
			<tr><td><a href="pl001005.png.html" target=_blank>
				<image src="pl001005t.png" alt="click to enlarge" align=right>
			</a></td></tr>
			<tr align=center><td><b>Fig. 5</b></td></tr>
		</table>
		<br clear=all>
		
		<li>
		Click <i>Remember</i> button. Every time you click this button, marked locations will be remembered with their assigned colors.
		<br clear=all>
		
		<li>
		Select a third color in <i>Use color</i> list, un-check <i>Live</i> box (i.e. ensure this option is un-selected), click user node "chr4", then nothing happens in the GenomeView window. This is because un-checked <i>Live</i> option lets GenomeView window not to response to user selections immediately.
		<br clear=all>
		
		<li>
		Ensure user node "chr4" is selected, mouse-right-click, select "Genome Drawing", then chr4 LOC P-nodes are marked by the third color.
		<br clear=all>
		
		<li>
		Click first <i>Clear</i> button (the one next to <i>Scale Up</i> button), then you will find that only chr4 LOC P-nodes are cleared. This is because chr2 and chr3 LOC P-nodes are remembered with their assigned colors.
		<br clear=all>
		
		<li>
		Check <i>Live</i> box, un-check <i>Additive</i> box, click map location user node "2p11.2", then you will find that corresponding locations are drawn twice in different colors.
		<table border=0 align=center>
			<tr><td><a href="pl001006.png.html" target=_blank>
				<image src="pl001006t.png" alt="click to enlarge" align=right>
			</a></td></tr>
			<tr align=center><td><b>Fig. 6</b></td></tr>
		</table>
		<br clear=all>
		
		<li>
		Double-click somewhere near 2p11.2 on chr2, an enlarged view window called "1 - GenomeView - Human" will appear. Note that a blue dash-line box will appear on chr2 for indicating the corresponding region of the enlarged view window. In the case that you have multiple enlarged view windows, the numbers appended to blue dash-line boxes and the numbers appended to their window titles will show the correspondence (see Fig. 7, blue and red circle).
		<table border=0 align=center>
			<tr><td><a href="pl001007.png.html" target=_blank>
				<image src="pl001007t.png" alt="click to enlarge" align=right>
			</a></td></tr>
			<tr align=center><td><b>Fig. 7</b></td></tr>
		</table>
		<br clear=all>
		
		<li>
		Move mouse cursor to the edge of enlarged view window, <i>drag</i> (i.e. hold mouse-lef-button and move mouse) to resize window. By widening enlarged view window, you will see the size of corresponding blue dash-line box is changed.
		<br clear=all>
		
		<li>
		The coordinate shows the chromosomeal position, doule-click on it to re-center the enlarged view (i.e. movements of enlarged view region). Try centering your enlarged view on chr2 to 87M base-pair.
		<br clear=all>
		
		<li>
		<i>Scale Down</i> or <i>Scale Up</i> so that you see chromosome from 84M to 90M.
		<br clear=all>
		
		<li>
		Drag mouse accross chromosome band and from 84M to 90M, a popup menu with two items (<i>label</i> and <i>de-label</i>) will appear (see Fig. 8), then select <i>label</i>. You will find that locations in dragged region will be labeled with their names, i.e. the node names of corresponding R-nodes (see Fig. 9. In this case, node names are gene IDs). For locations assigned with two colors, their labels will be followed by two colored squares. For locations assigned with one colors, their labels will be followed by one colored square.
		<table border=0 align=center>
			<tr>
				<td><a href="pl001008.png.html" target=_blank>
				<image src="pl001008t.png" alt="click to enlarge" align=right>
				</a></td>
				<td><a href="pl001009.png.html" target=_blank>
				<image src="pl001009t.png" alt="click to enlarge" align=right>
				</a></td>
			</tr>
			<tr align=center>
				<td><b>Fig. 8</b></td>
				<td><b>Fig. 9</b></td>
			</tr>
		</table>
		<br clear=all>
		
		<li>
		Move mouse cursor around the enlarged view window and stop for a while, a tool tip will appear.
			<ul>
			<li>Coordinate: "double click to re-center"
			<li>Location boxes: corresponding label name and "double-click to label"
			<li>Colored squares: corresponding information in user tree (the path infomation of this label, i.e. R-node) and "double-click to goto"
			</ul>
		Double-click on a colored square, then you will come back to GOBU window, and the corresponding R-node will become visible and selected. Note that if <i>Live</i> box is checked, this action will cause non-remembered locations to be cleared.
		<br clear=all>
		
		<li>
		Click second <i>Clear</i> button (the one next to <i>Remember</i> button), click first <i>Clear</i> button, then you will find that all LOC P-nodes are cleared. Second <i>Clear</i> button is for releasing all remembered locations.
		<br clear=all>
		
		<li>
		Check <i>Live</i> and un-check <i>Additive</i>, set <i>Appearance level</i> of user tree to <i>Representative</i>, then click on any GO node. The drawn locations should belong to <a href="../../operational/index.html#usertreeP2">selected</a> R-nodes.
		
		</ol>

		<table border=2 align=right><tr><td><a href="#top">top</a></td></tr></table>
		<h3><a name="tips">Special Tips</a></h3>
		<ol>
		<li>Check <i>Live</i> and <i>Additive</i>, set <i>Appearance level</i> of user tree to <i>Representative</i>, then select Search -&gt; Search GO tree. You may input your desired search pattern and options, you will see all related locations by clicking <i>Iterate all</i> button.
		</ol>

		<table border=2 align=right><tr><td><a href="#top">top</a></td></tr></table>
		<h3><a name="configTree">Config Tree</a></h3>
		<i>&lt;ItalicString&gt;</i> means a variable.<p>
		This plugin obtains genome names and corresponding chromosome lengths from <i><a href="../index.html#configTree">config tree</a></i>. Descendant <i>GenomeName</i> P-nodes of <i>Genome Information</i> P-node represent genomes. Under each <i>GenomeName</i> P-node, <a href="../../operational/index.html#userTree">normal</a> child nodes must in the form <i>&lt;chromosome name&gt;</i>:<i>&lt;chromosome length&gt;</i>.<p>
		See files "HumanChrInfo.tree", "MouseChrInfo.tree" and "ZebrafishChrInfo.tree" in <i>config</i> sub-folder for examples.

	</body>
</html>
